hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -343,6 +362,45 @@ func MachineDeploymentComplete(deployment *capiv1.MachineDeployment) bool {
 		newStatus.ObservedGeneration >= deployment.Generation
 }
 
+func generateMachineHealthCheck(nodePool *hyperv1.NodePool, targetNamespace, infraID string) *capiv1.MachineHealthCheck {
+	maxUnhealthy := intstr.FromInt(2)
+	resourcesName := generateName(infraID, nodePool.Spec.ClusterName, nodePool.GetName())
+	return &capiv1.MachineHealthCheck{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nodePool.GetName(),
+			Namespace: targetNamespace,
+		},
+		// Opinionated spec based on https://github.com/openshift/managed-cluster-config/blob/14d4255ec75dc263ffd3d897dfccc725cb2b7072/deploy/osd-machine-api/011-machine-api.srep-worker-healthcheck.MachineHealthCheck.yaml
+		// TODO (alberto): possibly expose this config at the nodePool API.
+		Spec: capiv1.MachineHealthCheckSpec{
```

Comment:
My read of the current API indicates what you're doing here is safe today, but it reminds me of a hazard with the approach of reconciling the entire object by returning the new spec state. If there are any fields on spec that are immutable or which are intended to be set by some other controller, this approach can cause errors by trying to stomp such fields (e.g. a Service). I [fixed one error like this](https://github.com/openshift/hypershift/pull/128/files#diff-2a286331d7ebff3abdaf711da406314a475a5fa6fb75535b03bcadf776881c5cL173) in the MCS code during the last refactor. The alternative approach is to just set fields you want to reconcile on the object spec explicitly rather than returning the entire spec like this.

Just a note for future reference. In this case it's fine but I wanted to share my experience.

Reply:
Make sense, I'll address this in the follow up refactor for this controller

Reply:
No problem, no changes necessary, just wanted to transmit the knowledge